src/dataframe/dataframe.jl

```diff
@@ -1223,34 +1223,90 @@ Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
+        throw(ArgumentError("`cols` keyword argument must be `:identical`, `:equal`, or `:intersect`"))
     end
     nrows, ncols = size(df)
+    targetrows = nrows + 1
     if ncols == 0 && row isa NamedTuple
         for (n, v) in pairs(row)
             setproperty!(df, n, fill!(Tables.allocatecolumn(typeof(v), 1), v))
         end
         return df
     end
-    # Only check for equal lengths, as an error will be thrown below if some names don't match
-    if columns === :equal && length(row) != ncols
-        # TODO: add tests for this case after the deprecation period
-        Base.depwarn("In the future push! will require that `row` has the same number" *
-                      " of elements as is the number of columns in `df`. " *
-                      "Use `columns=:intersect` to disable this check.", :push!)
+    if columns === :equal
+        # Only check for equal lengths if :equal is selected,
+        # as an error will be thrown below if some names don't match
+        if length(row) != ncols
+            # TODO: add tests for this case after the deprecation period
+            Base.depwarn("In the future push! will require that `row` has the same number" *
+                          " of elements as is the number of columns in `df`. " *
+                          "Use `columns=:intersect` to disable this check.", :push!)
+        end
+        if row isa NamedTuple
+            matching = all(x -> x[1]==x[2], zip(propertynames(row), _names(df)))
+            if !matching
+                Base.depwarn("columns=:equal as default is deprecated; in the future :identical " *
+                             "will be the default", :push!)
+            end
+        end
+    end
+    if columns === :identical
+        if length(row) != ncols
+            throw(ArgumentError("A pushed object must have the same number of elements" *
+                                " as a target data frame."))
+        end
+        if row isa NamedTuple
+            if any(x -> x[1] != x[2], zip(propertynames(row), _names(df)))
+                throw(ArgumentError("A pushed `NamedTuple` must have exactly the " *
+                                    "same column names and in the same order " *
+                                    "as a target data frame."))
+            end
+        end
+    end
+    if columns === :union
+        for k in keys(row)
+            if !hasproperty(df, k)
+                df[!, k] = Vector{Union{Missing, typeof(row[k])}}(missing, nrow(df))
+            end
+        end
     end
     current_col = 0
     try
         for (col, nm) in zip(_columns(df), _names(df))
             current_col += 1
-            push!(col, row[nm])
+            if columns === :union
+                val = get(row, nm, missing)
+                if val isa eltype(col)
+                    push!(col, val)
+                else
+                    newcol = similar(Vector{promote_type(typeof(val), eltype(col))},
```

Comment:
```suggestion
                    newcol = Tables.allocatecolumn(promote_type(typeof(val), eltype(col)),
```
(need to adjust indentation below)